File: Algorithms/060.PermutationSequence.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        string permutation;
        vector<int> s;
        for (int i=1; i <= n; i++) {
            s.push_back(i);
        }
        getPermutation(s, k, permutation);
        return permutation;
    }
private:
    int factorial(int n) {
        int result = 1;
        for (int i=2; i <= n; ++i) {
            result *= i;
        }
        return result;
    }
    void getPermutation(vector<int>& s, int k, string& permutation) {
        if (s.size() == 1) {
            permutation.push_back(s[0] + '0');
            return;
        }
        int p = factorial(s.size()-1);
        int a = s[(k-1)/p];
        permutation.push_back(a + '0');
        k -= ((k-1)/p)*p;
        s.erase(find(s.begin(), s.end(), a));
        getPermutation(s, k, permutation);
    }
};

File: Algorithms/py/060.PermutationSequence.py
class Solution:
    # @return a string
    def getPermutation(self, n, k):
        def _factorial(n):
            result = 1
            for i in range(2, n+1):
                result *= i
            return result

        def _kth_permutation(s, k, permutation):
            if len(s) == 1:
                permutation.append(str(s[0]))
                return
            p = _factorial(len(s)-1)
            a = s[(k-1)/p]
            permutation.append(str(a))
            k -= ((k-1)/p)*p
            s.remove(a)
            _kth_permutation(s, k, permutation)
        permutation = []
        s = range(1, n+1)
        _kth_permutation(s, k, permutation)
        return "".join(permutation)
